### PR TITLE
DimensionalData extension

### PR DESCRIPTION
**Pull request by longemen3000**
adds `DimensionaData` as an extension, closes https://github.com/rafaqz/DimensionalData.jl/pull/487

### CODE REVIEW

File: Project.toml [hunk 1]
```diff
@@ -19,8 +19,15 @@ OrderedCollections = "1"
...
 [extras]
+DimensionalData = "0703355e-b756-11e9-17c0-8b28908087d0"
 Test = "8dfed614-e22c-5e08-85e1-65c5234f0b40"
 
 [targets]
-test = ["Test"]
+test = ["DimensionalData", "Test"]
+
+[weakdeps]
+DimensionalData = "0703355e-b756-11e9-17c0-8b28908087d0"
```
Comment (odow) [inline Project.toml:33]:
```suggestion
DimensionalData = "0703355e-b756-11e9-17c0-8b28908087d0"

```

Comment (odow) [inline Project.toml:33]: So what happens with this in Julia 1.6? It just gets ignored?

Comment (longemen3000) [reply Project.toml:33]: yes

Comment (longemen3000) [reply Project.toml:33]: the options for Backwards compatibility is by `Requires` or as a regular dependency

File: Project.toml [hunk 2]
```diff
@@ -19,8 +19,15 @@ OrderedCollections = "1"
 SnoopPrecompile = "1"
 julia = "1.6"
 
+[extensions]
+JuMPDimensionalDataExt = "DimensionalData"
+
 [extras]
+DimensionalData = "0703355e-b756-11e9-17c0-8b28908087d0"
```
Comment (odow) [inline Project.toml:26]: TODO: add a [compat] entry

File: test/Containers/test_dimensionaldata.jl [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    @test_throws(
+        ArgumentError,
+        @variable(model, z[i = 1:3, j = i:2], container = DimArray),
+    )
+    @test_throws(
+        ArgumentError,
+        @variable(model, w[1:3, j = 1:2], container = DimArray),
+    )
+    return
+end
```
Comment (odow) [inline test/Containers/test_dimensionaldata.jl:29]:
```suggestion
end
```

File: Project.toml (hunk 1 above)
Comment (odow) [reply Project.toml:33]: I think I'm okay if this is a Julia 1.9+ feature only. We don't need backwards compatibility for new features.

One issue is that the docs are currently build using Julia 1.6. So we'd need to update that, and then add some big compat warnings.

**Comment by odow** [comment]
I guess a heads up: it's not certain that we'll merge an extension like this just yet. It's a pretty big change for JuMP, not because of this PR, but because it opens the door for others like it. But we have a monthly call on Thursday that I'll discuss this in.

See also: https://github.com/jump-dev/JuMP.jl/issues/1350

File: ext/JuMPDimensionalDataExt.jl [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+import DimensionalData
+import JuMP
+
+function JuMP.Containers.container(
+    f::F, 
+    indices::JuMP.Containers.VectorizedProductIterator,
+    c::Type{DimensionalData.DimArray},
+    names,
+) where {F}
```
Comment (odow) [inline ext/JuMPDimensionalDataExt.jl:16]:
```suggestion
) where {F<:Function}
```

File: test/Containers/test_dimensionaldata.jl [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    
+function test_dimension_data_integration()
+    if !isdefined(Base, :get_extension)
+        return
+    end
+    model = Model()
+    @variable(model, x[i = 2:4, j = ["a", "b"]], container = DimArray)
+    @test x isa DimArray
+    @test_throws(
+        ArgumentError,
```
Comment (odow) [inline test/Containers/test_dimensionaldata.jl:21]:
```suggestion
        MethodError,
```

File: ext/JuMPDimensionalDataExt.jl [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+        if i isa Symbol && !Base.isidentifier(i)
+            # A symbol was created automatically by JuMP.
+            throw(ArgumentError("Cannot build DimArray. invalid indices."))
+        end
+    end
+    dims = NamedTuple(i => j for (i, j) in zip(names, indices.prod.iterators))
+    return DimensionalData.DimArray(map(i -> f(i...), indices), dims)
+end
+
+# TODO: imporve error for nested iterators
```
Comment (odow) [inline ext/JuMPDimensionalDataExt.jl:27]:
```suggestion
# TODO: improve error for nested iterators
```

**Review by odow — COMMENTED** [review]
Overall, this is pretty slick.

Before merging, we need a bunch more tests, nicer error messages, and some documentation.

But I'm in favor of adding this to JuMP. The main issue I see is a discoverability one. We'll likely need some careful documentation about how and when different extensions are loaded and what features they enable.

File: test/Containers/test_dimensionaldata.jl [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+#  License, v. 2.0. If a copy of the MPL was not distributed with this
+#  file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+module TestContainersDimensionalData
+
+using Test
+
+using DimensionalData
+using JuMP
+
```
Comment (odow) [inline test/Containers/test_dimensionaldata.jl:12]:
```suggestion

```

File: test/Containers/test_dimensionaldata.jl [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+module TestContainersDimensionalData
+
+using Test
+
+using DimensionalData
+using JuMP
+    
+function test_dimension_data_integration()
+    if !isdefined(Base, :get_extension)
+        return
```
Comment (odow) [inline test/Containers/test_dimensionaldata.jl:15]:
```suggestion
        return  # Skip this test on Julia versions before v1.9
```

**Comment by odow** [comment]
Seems like this is working:

Julia 1.6
image

Julia 1.9
image

**Comment by odow** [comment]
Some notes from the JuMP call

 - [ ] What if extension makes a breaking release
 - [ ] Add documentation
 - [ ] Necessary criteria for acceptance
      - Bar for acceptance is high. Qualitative decision by JuMP developers
      - "Significant" package in the ecosystem
      - A useful extension for JuMP models
      - Mature: 1.0 release, _or_ battle tested for a long time
      - Self-contained

**Comment by longemen3000** [comment]
>  What if extension makes a breaking release
You can add `compat` annotations for extensions (but aqua seems to have problems with those, In general, Aqua needs to be updated to support 1.9 extensions and weakdeps).

**Comment by odow** [comment]
The compat issue was this:

 * What if DimensionalData releases a breaking `v0.25` that we don't want to (or can't) support? 
 * Are we stuck on v0.24 for ever? 
 * Can users install JuMP and v0.25 together and have the extension not load? 
 * Is it a breaking change if JuMP removes an extension?

**Comment by rafaqz** [comment]
This is an interesting point about extensions in general for the ecosystem... we are making our dependency trees larger and that does have costs besides compile time.

FWIW DimensionalData.jl is unlikely to do anything you cant support or change the basics of the dimarray interface much in the next few years.

**Comment by odow** [comment]
Based on my testing, it seems that the compat dependency issue is a hard requirement:
```julia
(dd) pkg> st
Status `/private/tmp/dd/Project.toml`
  [0703355e] DimensionalData v0.24.12
  [4076af6c] JuMP v1.11.1 `~/.julia/dev/JuMP`

(dd) pkg> add DimensionalData@0.23
   Resolving package versions...
ERROR: Unsatisfiable requirements detected for package DimensionalData [0703355e]:
 DimensionalData [0703355e] log:
 ├─possible versions are: 0.1.0-0.24.12 or uninstalled
 ├─restricted to versions 0.24 by JuMP [4076af6c], leaving only versions: 0.24.0-0.24.12 or uninstalled
 │ └─JuMP [4076af6c] log:
 │   ├─possible versions are: 1.11.1 or uninstalled
 │   └─JuMP [4076af6c] is fixed to version 1.11.1
 └─restricted to versions 0.23 by an explicit requirement — no versions left
```

Thus, I think we need to treat extensions as if we were adopting the dependency into JuMP proper. The lazy loading is almost an implementation detail, rather than an magic thing that gets loaded if the versions align.

This makes me a lot more hesitant to add extensions for packages which are not version 1, or which have large dependency trees.

In this case, DimensionalData is borderline, because it's pre-1.0, but it is fairly stable at v0.24.12, and it has a fairly reasonable set of dependencies: https://juliahub.com/ui/Packages/DimensionalData/Unppe/0.24.12?page=1

**Comment by odow** [comment]
Closing in favor of https://github.com/jump-dev/JuMP.jl/pull/3413

**Pull request closed without merging**